src/counts.jl

```diff
@@ -352,6 +351,9 @@ function addcounts_radixsort!(cm::Dict{T}, x::AbstractArray{T}) where T
     # It seems that sort is inferred in Julia 0.7.
     return _addcounts_radix_sort_loop!(cm, sx)
 end
+# general iterator form
+addcounts_radixsort!(cm::Dict{T}, x) where T = addcounts_radixsort!(cm, collect(x))
```

Comment:
That's fine, because you basically collect into the workspace imho.

Reply:
But this is going to make an unnecessary copy. I think the best solution is to make the method above accept any iterator, and do:
```julia
cx = x isa AbstractArray ? copy(x) : collect(x)
sx = sort!(cx, alg = RadixSort)
```
(`collect` sometimes returns an `Array` when `copy` would be more efficient, e.g. for `BitArray`)

Reply:
Is there an inplace radix sort?

Reply:
https://github.com/Moelf/StatsBase.jl/blob/6e40ec06a1532eaddf6b3e86e9a8c7e01e8aa323/src/counts.jl#L345

it doesn't make unnecessary copy because this is only fall back when `x` is not an array. so instead of doing that manual dispatch on `cx=`, this looks fine to me

Reply:
Yes actually it's [defined in terms of `sort!`](https://github.com/JuliaCollections/SortingAlgorithms.jl/blob/11e7bf999847b39a651db86d7bebca93612f4f50/src/SortingAlgorithms.jl#L64), and `sort` just calls `sort!` on a copy.

Reply:
When you call `addcounts_radixsort!(cm, collect(x))`, you allocate a new vector, and then inside `addcounts_radixsort!` `sort` makes another copy. This is redundant.